### PR TITLE
database/locker: Actually unlock advisory locks

### PR DESCRIPTION
**Pull request by efritz**
Apparently we were hiding error messages on unlock. I'm not sure if there was a recent change or if this has just been a problem for a while, but it may be affecting the lag between an LSIF upload and it being available for queries.

*Also* apparently we were not always calling Lock/Unlock in the same session (making the Unlock fail _silently until now_). Holding a lock within a transaction stops the connection from being broken and ensures that the unlock happens in the same session.

In order to ensure this happens, we split the lock method into `Lock` and `LockInTransaction` that errors when you call with the incorrect txn state.

### CODE REVIEW

**Review by eseliger — APPROVED** [review]
Can we test this in any way?

File: internal/database/locker/locker.go [hunk 1]
```diff
@@ -3,6 +3,7 @@ package locker
 import (
 	"context"
 	"database/sql"
+	"errors"
```
[the remark this answers is not in the document]
Comment (efritz) [reply internal/database/locker/locker.go:6]: 👊

[the remark this answers is not in the document]
Comment (efritz) [reply internal/database/locker/locker.go:6]: Yes

File: internal/database/locker/locker.go [hunk 2]
```diff
@@ -107,10 +113,20 @@ const tryLockQuery = `
 SELECT pg_try_advisory_lock(%s, %s)
 `
 
+var ErrUnlock = errors.New("Failed to unlock")
```
Comment (efritz) [inline internal/database/locker/locker.go:116]:
```suggestion
var ErrUnlock = errors.New("failed to unlock")
```

File: internal/database/locker/locker.go [hunk 3]
```diff
@@ -53,64 +54,87 @@ func (l *Locker) Transact(ctx context.Context) (*Locker, error) {
 // occurs during unlock, the error is added to the resulting error value.
 type UnlockFunc func(err error) error
 
+// ErrNoTransaction occurs when Lock is called outside of a transaction.
+var ErrNoTransaction = errors.New("store: not in a transaction")
```
Comment (efritz) [inline internal/database/locker/locker.go:58]:
```suggestion
var ErrNoTransaction = errors.New("locker: not in a transaction")
```

File: internal/database/locker/locker.go [hunk 4]
```diff
@@ -53,64 +54,87 @@ func (l *Locker) Transact(ctx context.Context) (*Locker, error) {
...
-		locked, err = l.lock(ctx, key)
+		locked, err = l.selectAdvisoryLock(ctx, key)
 	} else {
-		locked, err = l.tryLock(ctx, key)
+		locked, err = l.selectTryAdvisoryLock(ctx, key)
 	}
 
 	if err != nil || !locked {
-		return false, nil, err
+		return false, nil, l.Done(err)
```
Comment (efritz) [inline internal/database/locker/locker.go:80]:
```suggestion
		return false, nil, err
```

File: internal/database/locker/locker.go [hunk 5]
```diff
@@ -53,64 +54,87 @@ func (l *Locker) Transact(ctx context.Context) (*Locker, error) {
...
 	}
 
 	unlock := func(err error) error {
-		if unlockErr := l.unlock(key); unlockErr != nil {
+		if unlockErr := l.unlock(context.Background(), key); unlockErr != nil {
 			err = multierror.Append(err, unlockErr)
 		}
 
-		return err
+		return l.Done(err)
```
Comment (efritz) [inline internal/database/locker/locker.go:88]:
```suggestion
		return err
```

**Comment by efritz** [comment]
@eseliger Actually is very hard to test as we need distinct connections to verify blocking behavior.

File: internal/database/locker/locker_test.go [hunk 6]
```diff
@@ -110,3 +110,27 @@ func TestLockBlockingAcquire(t *testing.T) {
 		t.Errorf("lock not acquired before release")
 	}
 }
+
+func TestLockBadTransactionState(t *testing.T) {
+	if testing.Short() {
+		t.Skip()
+	}
+	db := dbtesting.GetDB(t)
```
Comment (eseliger) [inline internal/database/locker/locker_test.go:118]: Just tryna advocate for using the "new way" ™️, feel free to ignore

```suggestion
	db := dbtest.NewDB(t, "")
```

**Review by eseliger — APPROVED** [review]
I must say this is pretty complex(not your fault), but I hope I got it now. 
LGTM! Nice work 🌟

**Pull request merged**